pkg/openapi/validation.go

```diff
@@ -0,0 +1,263 @@
+package openapi
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/nirmata/kyverno/data"
+
+	"github.com/golang/glog"
+
+	"github.com/nirmata/kyverno/pkg/engine"
+	"github.com/nirmata/kyverno/pkg/engine/context"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	v1 "github.com/nirmata/kyverno/pkg/api/kyverno/v1"
+
+	openapi_v2 "github.com/googleapis/gnostic/OpenAPIv2"
+	"github.com/googleapis/gnostic/compiler"
+	"k8s.io/kube-openapi/pkg/util/proto"
+	"k8s.io/kube-openapi/pkg/util/proto/validation"
+
+	"gopkg.in/yaml.v2"
+)
+
+var openApiGlobalState struct {
+	document             *openapi_v2.Document
+	definitions          map[string]*openapi_v2.Schema
+	kindToDefinitionName map[string]string
+	models               proto.Models
+	isSet                bool
+}
+
+func init() {
+	err := setValidationGlobalState()
+	if err != nil {
+		panic(err)
+	}
+}
+
+func UseCustomOpenApiDocument(customDoc []byte) error {
+	var spec yaml.MapSlice
+	err := yaml.Unmarshal(customDoc, &spec)
+	if err != nil {
+		return err
+	}
+
+	openApiGlobalState.document, err = openapi_v2.NewDocument(spec, compiler.NewContext("$root", nil))
+	if err != nil {
+		return err
+	}
+
+	openApiGlobalState.definitions = make(map[string]*openapi_v2.Schema)
+	openApiGlobalState.kindToDefinitionName = make(map[string]string)
+	for _, definition := range openApiGlobalState.document.GetDefinitions().AdditionalProperties {
+		openApiGlobalState.definitions[definition.GetName()] = definition.GetValue()
+		path := strings.Split(definition.GetName(), ".")
+		openApiGlobalState.kindToDefinitionName[path[len(path)-1]] = definition.GetName()
+	}
+
+	openApiGlobalState.models, err = proto.NewOpenAPIData(openApiGlobalState.document)
+	if err != nil {
+		return err
+	}
+
+	openApiGlobalState.isSet = true
+
+	return nil
+}
+
+func ValidatePolicyMutation(policy v1.ClusterPolicy) error {
+	if !openApiGlobalState.isSet {
+		glog.V(4).Info("Cannot Validate policy: Validation global state not set")
+		return nil
+	}
+
+	var kindToRules = make(map[string][]v1.Rule)
+	for _, rule := range policy.Spec.Rules {
+		rule.MatchResources.Selector = nil
```

Comment:
We want to apply the mutation on the empty resource, having any kind of value in the match block would make sure the empty resource would not get mutated.

We would need to make everything nil except rule.matchresource.kinds. Also since kinds is being made optional validate policy needs to be refactored.